File: src/main/nodejs/sample/ProcessMessageExchangeClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessOperation = void 0;
const MESSAGE_DELIMITER = "__DEL__";
const MESSAGE_END_OF_PROCESS = "__EOP__";
function ProcessOperation(callBack) {
    let buffer = ''; // Buffer to collect incoming data
    let serverRequest = '';
    // Listen for data from the parent process via stdin
    process.stdin.on('data', (data) => {
        buffer += data.toString(); // Collect incoming data
        // Check if the received data contains newline characters
        if (buffer.includes('\n')) {
            // Split the data into lines
            const lines = buffer.split('\n');
            // Process each line
            for (let i = 0; i < lines.length - 1; i++) {
                const line = lines[i];
                if (line.includes(MESSAGE_DELIMITER)) {
                    let response = callBack(serverRequest);
                    response += MESSAGE_DELIMITER;
                    console.log(response);
                    serverRequest = '';
                }
                else if (!line.includes(MESSAGE_END_OF_PROCESS)) {
                    serverRequest += line + "\n";
                    //console.log('Child received message from parent:', line);
                }
                else if (buffer.includes(MESSAGE_END_OF_PROCESS)) {
                    process.exit(0);
                }
            }
            // Update the buffer with any remaining data
            buffer = lines[lines.length - 1];
        }
    });
    // Keep the process alive by running an empty loop
    setInterval(() => { }, 1000); // Adjust the interval as needed
}
exports.ProcessOperation = ProcessOperation;

File: src/main/python/sample/ProcessMessageExchangeClient.py
# ProcessMessageExchangeClient.py

# Define constants
MESSAGE_DELIMITER = "__DEL__"
MESSAGE_END_OF_PROCESS = "__EOP__"

# Function to get a request from the user (stdin)
def get_server_request():
    return input()

# Function to reply with a response and delimiter
def reply_response(response):
    print(response)
    print(MESSAGE_DELIMITER)

# Function to process messages
def process_message(process_request):
    try:
        message = ""

        while True:
            input_data = get_server_request()

            if input_data is None:
                break

            if input_data == MESSAGE_END_OF_PROCESS:
                break

            if input_data == MESSAGE_DELIMITER:
                message = process_request(message)
                reply_response(message)
                message = ""
                continue

            if message == "":
                message = input_data
            else:
                message += f"\n{input_data}"

    except Exception as e:
        print(e)

# Call the main message processing function
if __name__ == "__main__":
    process_message(None)
